File: shell/preload/sitecustomize.js
const Module = require('module');
const path = require('path');
const client = require('./client.js');
const scopedEnvironment = require('./scoped-env.js');
scopedEnvironment.apply();

// 注册 ESM loader，使全局安装的包也可通过 import 导入
try {
  Module.register(new URL('esm-loader.mjs', `file://${__dirname}/`).href);
} catch (_) {}

function preferGlobalNodeModules() {
  const { QL_NODE_GLOBAL_PATH } = process.env;
  if (!QL_NODE_GLOBAL_PATH || Module._qlGlobalPathPatched) {
    return;
  }

  const originalResolveFilename = Module._resolveFilename;
  Module._resolveFilename = function (request, parent, isMain, options) {
    if (
      !Module.builtinModules.includes(request) &&
      !request.startsWith('node:') &&
      !request.startsWith('.') &&
      !path.isAbsolute(request)
    ) {
      try {
        return originalResolveFilename.call(this, request, parent, isMain, {
          ...options,
          paths: [QL_NODE_GLOBAL_PATH],
        });
      } catch (error) {}
    }

    return originalResolveFilename.call(this, request, parent, isMain, options);
  };
  Module._qlGlobalPathPatched = true;
}

function expandRange(rangeStr, max) {
  const tempRangeStr = rangeStr
    .trim()
    .replace(/-max/g, `-${max}`)
    .replace(/max-/g, `${max}-`);

  return tempRangeStr.split(' ').flatMap((part) => {
    const rangeMatch = part.match(/^(\d+)([-~_])(\d+)$/);
    if (rangeMatch) {
      const [, start, , end] = rangeMatch.map(Number);
      const step = start < end ? 1 : -1;
      return Array.from(
        { length: Math.abs(end - start) + 1 },
        (_, i) => start + i * step,
      );
    }
    return Number(part);
  });
}

function run() {
  const { envParam, numParam, PREV_NODE_OPTIONS } = process.env;
  process.env.NODE_OPTIONS = PREV_NODE_OPTIONS || '';

  if (envParam && numParam) {
    const array = (process.env[envParam] || '').split('&');
    const runArr = expandRange(numParam, array.length);
    const arrayRun = runArr.map((i) => array[i - 1]);
    const envStr = arrayRun.join('&');
    process.env[envParam] = envStr;
  }
}

try {
  if (!process.argv[1]) {
    return;
  }

  preferGlobalNodeModules();

  process.on('SIGTERM', (code) => {
    process.exit(15);
  });

  run();

  const { sendNotify } = require('./__ql_notify__.js');
  global.QLAPI = {
    notify: sendNotify,
    ...client,
  };
} catch (error) {
  console.log(`run builtin code error: `, error, '\n');
}

File: shell/preload/sitecustomize.py
import os
import re
import builtins
import sys
import scoped_env
scoped_env.apply_scoped_environment()
import signal
from client import Client


def try_parse_int(value):
    try:
        return int(value)
    except ValueError:
        return None


def expand_range(range_str, max_value):
    temp_range_str = (
        range_str.strip()
        .replace("-max", f"-{max_value}")
        .replace("max-", f"{max_value}-")
    )

    result = []
    for part in temp_range_str.split(" "):
        range_match = re.match(r"^(\d+)([-~_])(\d+)$", part)
        if range_match:
            start, _, end = map(try_parse_int, range_match.groups())
            step = 1 if start < end else -1
            result.extend(range(start, end + step, step))
        else:
            result.append(int(part))

    return result


def run():
    os.environ["PYTHONPATH"] = os.getenv("PREV_PYTHONPATH", "")

    env_param = os.getenv("envParam")
    num_param = os.getenv("numParam")

    if env_param and num_param:
        array = (os.getenv(env_param) or "").split("&")
        run_arr = expand_range(num_param, len(array))
        array_run = [array[i - 1] for i in run_arr if i - 1 < len(array) and i > 0]
        env_str = "&".join(array_run)
        os.environ[env_param] = env_str


def handle_sigterm(signum, frame):
    sys.exit(15)


try:
    signal.signal(signal.SIGTERM, handle_sigterm)

    run()

    from __ql_notify__ import send

    class BaseApi(Client):
        def notify(self, *args, **kwargs):
            return send(*args, **kwargs)

    QLAPI = BaseApi()
    builtins.QLAPI = QLAPI
except Exception as error:
    print(f"run builtin code error: {error}\n")
